File: Algorithms/src/1411. Number of Ways to Paint N × 3 Grid/Code/solution.py
class Solution:
    def numOfWays(self, n: int) -> int:
        MOD = 10**9 + 7

        same = 6  # ABA
        diff = 6  # ABC

        for _ in range(2, n + 1):
            new_same = (same * 3 + diff * 2) % MOD
            new_diff = (same * 2 + diff * 2) % MOD

            same, diff = new_same, new_diff

        return (same + diff) % MOD

File: Algorithms/src/1411. Number of Ways to Paint N × 3 Grid/Code/solution.java
class Solution {
    public int numOfWays(int n) {
        final int MOD = 1_000_000_007;

        long same = 6;
        long diff = 6;

        for (int i = 2; i <= n; i++) {
            long newSame = (same * 3 + diff * 2) % MOD;
            long newDiff = (same * 2 + diff * 2) % MOD;

            same = newSame;
            diff = newDiff;
        }

        return (int) ((same + diff) % MOD);
    }
}
